### PR TITLE
Add amp-bind optimizer

### PR DESCRIPTION
**Pull request by jridgewell**
See https://github.com/ampproject/amphtml/pull/32851

Part of https://github.com/ampproject/amphtml/issues/27590

### CODE REVIEW

File: packages/optimizer/lib/DomTransformer.js [hunk 1]
```diff
@@ -36,6 +36,8 @@ const TRANSFORMATIONS_AMP_FIRST = [
   'OptimizeImages',
   // Detect hero image and preload link rel=preload, needs to run after OptimizeImages
   'OptimizeHeroImages',
+  // Inject a querySelectorAll query-able i-amphtml-binding attribute on elements with bindings.
```
Comment (sebastianbenz) [inline packages/optimizer/lib/DomTransformer.js:39]:
```suggestion
  // Inject a querySelectorAll query-able i-amphtml-binding attribute on elements with bindings.
  // This needs to run after AutoExtensionImporter.
```

File: packages/optimizer/lib/DomTransformer.js [hunk 2]
```diff
@@ -73,6 +75,8 @@ const TRANSFORMATIONS_PAIRED_AMP = [
   'OptimizeImages',
   // Detect hero image and preload link rel=preload
   'OptimizeHeroImages',
+  // Inject a querySelectorAll query-able i-amphtml-binding attribute on elements with bindings.
```
Comment (sebastianbenz) [inline packages/optimizer/lib/DomTransformer.js:78]:
```suggestion
  // Inject a querySelectorAll query-able i-amphtml-binding attribute on elements with bindings.
  // This needs to run after AutoExtensionImporter.
```

File: packages/optimizer/lib/transformers/OptimizeAmpBind.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+const {skipNodeAndChildren} = require('../HtmlDomHelper');
+const {isTemplate} = require('../AmpConstants');
+
+/**
+ */
+class OptimizeAmpBind {
+  constructor(config) {
+    this.log_ = config.log.tag('OptimizeAmpBind');
+
+    // setup implementation only if placeholder generation is enabled
```
Comment (sebastianbenz) [inline packages/optimizer/lib/transformers/OptimizeAmpBind.js:27]: Looks like a leftover from a different transformer

```suggestion
```

File: packages/optimizer/lib/transformers/OptimizeAmpBind.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    this.enabled_ = !!config.optimizeAmpBind;
```
Comment (sebastianbenz) [inline packages/optimizer/lib/transformers/OptimizeAmpBind.js:28]: We should enable this by default.

```suggestion
    this.enabled_ = config.optimizeAmpBind !== false;
```

Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeAmpBind.js:28]: Or did you intentionally leave this disabled until the validator changes are live?

**Review by sebastianbenz — CHANGES_REQUESTED** [review]
Nice! 

Two comments:

- using `data-amphtml-binding` instead of `I-amphtml-binding` would mean that we wouldn't have to update the validator
- can you please add a test here: https://github.com/ampproject/amp-toolbox/tree/main/packages/optimizer/spec/transformers/valid. You can run `npm run optimier:test:snapshot` to generate the expected output files

**Comment by sebastianbenz** [comment]
//cc @westonruter @schlessera for porting this to the PHP optimizer.

File: packages/optimizer/lib/transformers/OptimizeAmpBind.js (hunk 4 above)
Comment (jridgewell) [reply packages/optimizer/lib/transformers/OptimizeAmpBind.js:27]: Yup, this is intentionally disabled until validator changes are live in AMP Cache. I was gonna follow up with a PR to enable it afterwards.

**Review by jridgewell — COMMENTED** [review]
> - using data-amphtml-binding instead of I-amphtml-binding would mean that we wouldn't have to update the validator

I actually chose not to do this, since the attribute can only be valid in transformed outputs. Data attributes are always valid.

> - can you please add a test here: https://github.com/ampproject/amp-toolbox/tree/main/packages/optimizer/spec/transformers/valid. You can run npm run optimier:test:snapshot to generate the expected output files

Working on that now.

**Comment by jridgewell** [comment]
Tests added

**Review by sebastianbenz — APPROVED** [review]
Thanks!

**Pull request merged**